### PR TITLE
Assert no serialization errors in tests

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today in tests we handle deserialization errors much as in production
code, dropping the message and logging a warning. Tests typically fail
if they hit such an error, but there might be many retries first and and
the failure might just be reported as a timeout which is hard to
diagnose. Some tests might even still pass if, for instance, the bad
message is something like a background stats request.

With this commit we fail tests immediately on the first deserialization
problem they encounter.

Relates #81110

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java [hunk 1]
```diff
@@ -325,13 +338,17 @@ private static void sendErrorResponse(String actionName, TransportChannel transp
...
-            logger.warn(new ParameterizedMessage("Failed to deserialize response from [{}]", remoteAddress), serializationException);
-            handleException(handler, serializationException);
-            return;
+            try {
+                final Exception serializationException = new TransportSerializationException(
+                    "Failed to deserialize response from handler [" + handler + "]",
+                    e
+                );
+                logger.warn(new ParameterizedMessage("Failed to deserialize response from [{}]", remoteAddress), serializationException);
+                handleException(handler, serializationException);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/transport/InboundHandler.java:347]: can we not avoid the try-catch by just asserting before calling `handleException`?

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java [hunk 2]
```diff
@@ -361,15 +378,21 @@ private static void sendErrorResponse(String actionName, TransportChannel transp
...
+        boolean readSuccess = false;
         try {
             error = stream.readException();
+            readSuccess = true;
         } catch (Exception e) {
             error = new TransportSerializationException(
                 "Failed to deserialize exception response from stream for handler [" + handler + "]",
                 e
             );
         }
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/transport/InboundHandler.java:390]: Why not just `ignoreDeserializationErrors : error` inside the catch block? Seems I am missing the point here :slightly_smiling_face:

**Review by henningandersen — APPROVED** [review]
LGTM.

**Review by DaveCTurner — COMMENTED** [review]
For some reason I wanted to run `handleError` before tripping the assertion, but I've no idea why now. Simplified as you suggested.

**Pull request merged**